x/gamm/keeper/pool_service.go

```diff
@@ -296,25 +295,14 @@ func (k Keeper) JoinSwapShareAmountOut(
 		return sdk.Int{}, err
 	}
 
-	panic("implement") // I moved this past return, it caused everything beneath it to be dead code
-
-	tokenInAmount = sdk.ZeroInt()
-
-	// normalizedWeight := PoolAsset.Weight.ToDec().Quo(pool.GetTotalWeight().ToDec())
-	// tokenInAmount = calcSingleInGivenPoolOut(
-	// 	PoolAsset.Token.Amount.ToDec(),
-	// 	normalizedWeight,
-	// 	pool.GetTotalShares().Amount.ToDec(),
-	// 	shareOutAmount.ToDec(),
-	// 	pool.GetPoolSwapFee(ctx),
-	// ).TruncateInt()
-
-	if tokenInAmount.LTE(sdk.ZeroInt()) {
-		return sdk.Int{}, sdkerrors.Wrapf(types.ErrInvalidMathApprox, "token amount is zero or negative")
+	extendedPool, ok := pool.(types.PoolAmountOutExtension)
+	if !ok {
+		return sdk.Int{}, fmt.Errorf("pool with id %d does not support this kind of join", poolId)
 	}
 
-	if tokenInAmount.GT(tokenInMaxAmount) {
-		return sdk.Int{}, sdkerrors.Wrapf(types.ErrLimitMaxAmount, "%s token is larger than max amount", tokenInDenom)
+	tokenInAmount, err = extendedPool.JoinSwapShareAmountOut(ctx, tokenInDenom, shareOutAmount, tokenInMaxAmount)
```

Comment:
We should add the `tokenInAmount < tokenInMaxAmount` check at the keeper layer here